overlord/configstate/configcore/vitality.go

```diff
@@ -34,13 +39,79 @@ func init() {
 	supportedConfigurations["core."+vitalityOpt] = true
 }
 
-func handleVitalityConfiguration(tr config.ConfGetter, opts *fsOnlyContext) error {
-	// XXX: rewrite/restart changed configs
+func isNotInstalled(err error) bool {
+	_, ok := err.(*snap.NotInstalledError)
+	return ok
+}
+
+func handleVitalityConfiguration(tr config.Conf, opts *fsOnlyContext) error {
+	var pristineVitalityStr, newVitalityStr string
+
+	if err := tr.GetPristine("core", vitalityOpt, &pristineVitalityStr); err != nil && !config.IsNoOption(err) {
+		return err
+	}
+	if err := tr.Get("core", vitalityOpt, &newVitalityStr); err != nil && !config.IsNoOption(err) {
+		return err
+	}
+	if pristineVitalityStr == newVitalityStr {
+		return nil
+	}
+
+	st := tr.State()
+	st.Lock()
+	defer st.Unlock()
+
+	oldVitalityMap := map[string]int{}
+	newVitalityMap := map[string]int{}
+	// assign "0" (delete) rank to all old entries
+	for i, instanceName := range strings.Split(pristineVitalityStr, ",") {
+		oldVitalityMap[instanceName] = i + 1
+		newVitalityMap[instanceName] = 0
+	}
+	// build rank of the new entries
+	for i, instanceName := range strings.Split(newVitalityStr, ",") {
+		newVitalityMap[instanceName] = i + 1
+	}
+
+	for instanceName, rank := range newVitalityMap {
+		info, err := snapstate.CurrentInfo(st, instanceName)
+		if isNotInstalled(err) {
+			continue
+		}
+		if err != nil {
+			return err
+		}
+		// nothing rank to do if rank is unchanged
+		if oldVitalityMap[instanceName] == newVitalityMap[instanceName] {
+			continue
+		}
+		// rank changed, rewrite/restart services
+		for _, app := range info.Apps {
```

Comment:
This will unconditionally enable and start all services of the snap, even those previously disabled by the user. Is this expected outcome?

Reply:
I fixed this now but the fix is a bit dodgy and probably needs some tweaking and most likely a separate PR. The whole wrappers package is now a bit hard to use and there are too many indirections from "snapstaet->backend->wrappers" but I guess that is a separate discussion. I also added a spread test so that we can ensure this works correctly.

Reply:
Fair enough, I'd actually prefer a "conservative" fix as anything more could clash/complicate changes to services I've been working on.